downloader/src/main/kotlin/vcs/GitWorkingTree.kt

```diff
@@ -21,68 +21,85 @@ package com.here.ort.downloader.vcs
 
 import com.here.ort.downloader.WorkingTree
 import com.here.ort.model.VcsInfo
-import com.here.ort.utils.ProcessCapture
 
 import java.io.File
 
-private const val GIT_STATUS_BRANCH_UPSTREAM_PREFIX = "# branch.upstream "
+import org.eclipse.jgit.api.LsRemoteCommand
+import org.eclipse.jgit.lib.BranchConfig
+import org.eclipse.jgit.lib.Constants
+import org.eclipse.jgit.lib.Repository
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder
+import org.eclipse.jgit.submodule.SubmoduleWalk
 
 open class GitWorkingTree(workingDir: File, private val gitBase: GitBase) : WorkingTree(workingDir, gitBase.type) {
-    override fun isValid(): Boolean {
-        if (!workingDir.isDirectory) {
-            return false
+    private val repo = FileRepositoryBuilder().readEnvironment().findGitDir(workingDir).runCatching {
+        build()
+    }.getOrNull()
+
+    override fun isValid(): Boolean = repo?.objectDatabase?.exists() == true
+
+    override fun isShallow(): Boolean = repo?.directory?.resolve("shallow")?.isFile == true
+
+    private fun listSubmodulePaths(repo: Repository): List<String> {
+        fun listSubmodules(parent: String, repo: Repository, paths: MutableList<String>) {
+            val prefix = if (parent.isEmpty()) parent else "$parent/"
+
+            SubmoduleWalk.forIndex(repo).use { walk ->
+                while (walk.next()) {
+                    paths += "$prefix${walk.path}"
+                    walk.repository.use { submoduleRepo ->
+                        listSubmodules(paths.last(), submoduleRepo, paths)
+                    }
+                }
+            }
         }
 
-        // Do not use runGitCommand() here as we do not require the command to succeed.
-        val isInsideWorkTree = ProcessCapture(workingDir, "git", "rev-parse", "--is-inside-work-tree")
-        return isInsideWorkTree.isSuccess && isInsideWorkTree.stdout.trimEnd().toBoolean()
+        return mutableListOf<String>().also { paths ->
+            listSubmodules("", repo, paths)
+        }
     }
 
-    override fun isShallow(): Boolean {
-        val dotGitDir = gitBase.run(workingDir, "rev-parse", "--absolute-git-dir").stdout.trimEnd()
-        return File(dotGitDir, "shallow").isFile
-    }
+    override fun getNested(): Map<String, VcsInfo> =
+        runCatching {
+            // This throws a NullPointerException if "repo" is null.
```

Comment:
Is that comment necessary? Because it's basically the definition of `!!`.

Reply:
Hmm, I don't believe that's the definition of `!!`. That's just a cast, and if the code using `repo` would handle the `null`-case explicitly no exception would be thrown. So my "This" refers to the `listSubmodulePaths` function, not to the `!!` cast, and I wanted to express that the former throws a `NullPointerException` if `repo` is `null`, but we don't need to explicitly check for `repo == null` beforehand, as the code is wrapped into `runCatching` anyway.

So this also is a late handling of the error case when `repo` was initialized to `null` in line 35.

Reply:
`!!` is not only a cast, but also throws an NPE if the object is null, if the code would handle the `null`-case separately `!!` would not be needed. I thought it's obvious and the comment is superfluous. But if the NPE is the only expected exception, wouldn't a normal `if(repo == null) ...` be better than using `runCatching`?

Reply:
> `!!` is not only a cast, but also throws an NPE if the object is null

Doh, you're [right](https://kotlinlang.org/docs/reference/null-safety.html#the--operator), I somehow remembered this to only be a hint to the compiler...

In that case I can simply omit the `!!` and let the JGit methods throw the NPE if `repo` is `null`.